# README 


generate_fasta_4_haplotype.R: this script was used to generate the fasta files for each of the target genes. Sample IDs were used as the sequence IDs. Strings "_maj" and "_min" were included as the suffix of the IDs for major and minor haplotypes, respectively. 

Based on the metadata of the samples, the sequences in the fasta files were seperated to seperate files based on their courtry origin. Sequences with 'N' were excluded. Only major haplotypes were included. 

vaxpack_in_out_args.R: this R script was developed so that we can perform vaxpack analysis by providing comand line arguments.



Reference

https://github.com/myonaung/Naung-et-al-2021
 
# Pf_capture_div_code
