# README 


The script `generate_fasta_4_haplotype.R` was utilized to create fasta files for each of the target genes, with sample IDs used as sequence IDs. The IDs for major and minor haplotypes were differentiated by adding "_maj" and "_min" as suffixes, respectively. 

Using metadata from the samples, the fasta sequences were separated into individual files based on their country of origin, excluding sequences with 'N'. Only major haplotypes were included in the generated files.

The R script `vaxpack_in_out_args.R` was developed to enable vaxpack analysis through command line arguments.



